File: src/dynamic programming/cpp/all construct/test/test_all_construct.cpp
#include "all_construct.h"
#include "gtest/gtest.h"
#include <algorithm>

bool equal(std::vector<std::vector<std::string>> &v1,
           std::vector<std::vector<std::string>> &v2) {
  auto _equal = [](std::vector<std::string> &v1, std::vector<std::string> &v2) {
    std::sort(v1.begin(), v1.end());
    std::sort(v2.begin(), v2.end());
    return v1 == v2;
  };

  if (v1.size() != v2.size())
    return false;

  if (v1.empty())
    return true;

  std::sort(v1.begin(), v1.end());
  std::sort(v2.begin(), v2.end());

  for (int i = 0; i < v1.size(); i++) {
    if (!_equal(v1[i], v2[i]))
      return false;
  }

  return true;
}

TEST(AllConstructTest, TestNegativeBasic1) {
  std::vector<std::string> wordBank = {"bo",  "rd", "ate", "t",
                                       "ska", "sk", "boar"};
  std::string target = "skateboard";
  auto result = allConstructBasic(target, wordBank);

  ASSERT_TRUE(result.empty());
}

File: src/dynamic_programming/python/all_construct/tests/test_all_construct.py
import unittest

from src.all_construct import (
    all_construct_basic,
    all_construct_memo,
    all_construct_table,
)


def compare_2d_lists(l1, l2):
    if len(l1) != len(l2):
        return False

    for sublist1, sublist2 in zip(l1, l2):
        if set(sublist1) != set(sublist2):
            return False

    return True


class TestAllConstructBasic(unittest.TestCase):
    def test_no_combinations_possible(self):
        word_bank = ["bo", "rd", "ate", "t", "ska", "sk", "boar"]
        target = "skateboard"
        result = all_construct_basic(target, word_bank)
        self.assertEqual(result, [])

    def test_no_combinations_possible_2(self):
        word_bank = ["mo", "ha", "cz"]
        target = "mocha"
        result = all_construct_basic(target, word_bank)
        self.assertEqual(result, [])

    def test_single_combination_possible(self):
        word_bank = ["a", "b", "c"]
        target = "abc"
        result = all_construct_basic(target, word_bank)
        self.assertTrue(compare_2d_lists(result, [["a", "b", "c"]]))

    def test_multiple_combinations_possible(self):
        word_bank = ["ab", "abc", "cd", "def", "abcd"]
        target = "abcdef"
        result = all_construct_basic(target, word_bank)
        self.assertTrue(compare_2d_lists(result, [["abc", "def"]]))


class TestAllConstructMemo(unittest.TestCase):
    def test_no_combinations_possible(self):
        word_bank = ["bo", "rd", "ate", "t", "ska", "sk", "boar"]
        target = "skateboard"
        result = all_construct_memo(target, word_bank)
        self.assertEqual(result, [])

    def test_no_combinations_possible_2(self):
        word_bank = ["mo", "ha", "cz"]
        target = "mocha"
        result = all_construct_memo(target, word_bank)
        self.assertEqual(result, [])

    def test_single_combination_possible(self):
        word_bank = ["a", "b", "c"]
        target = "abc"
        result = all_construct_memo(target, word_bank)
        self.assertTrue(compare_2d_lists(result, [["a", "b", "c"]]))

    def test_multiple_combinations_possible(self):
        word_bank = ["ab", "abc", "cd", "def", "abcd"]
        target = "abcdef"
        result = all_construct_memo(target, word_bank)
        self.assertTrue(compare_2d_lists(result, [["abc", "def"]]))


class TestAllConstructTab(unittest.TestCase):
    def test_no_combinations_possible(self):
        word_bank = ["bo", "rd", "ate", "t", "ska", "sk", "boar"]
        target = "skateboard"
        result = all_construct_table(target, word_bank)
        self.assertEqual(result, [])

    def test_no_combinations_possible_2(self):
        word_bank = ["mo", "ha", "cz"]
        target = "mocha"
        result = all_construct_table(target, word_bank)
        self.assertEqual(result, [])

    def test_single_combination_possible(self):
        word_bank = ["a", "b", "c"]
        target = "abc"
        result = all_construct_table(target, word_bank)
        self.assertTrue(compare_2d_lists(result, [["a", "b", "c"]]))

    def test_multiple_combinations_possible(self):
        word_bank = ["ab", "abc", "cd", "def", "abcd"]
        target = "abcdef"
        result = all_construct_table(target, word_bank)
        self.assertTrue(compare_2d_lists(result, [["abc", "def"]]))


if __name__ == "__main__":
    unittest.main()
